external/llvm-project/mlir/include/mlir/Dialect/AMDGPU/IR/AMDGPU.td

```diff
@@ -411,6 +412,47 @@ def AMDGPU_RawBufferAtomicUminOp :
   let hasVerifier = 1;
 }
 
+def AMDGPU_DPP : AMDGPU_Op<"dpp", [SameTypeOperands, AllTypesMatch<["result", "src"]>]>,
+  Arguments<(ins AnyType:$src,
+                  OptionalAttr<ArrayAttr>:$quad_perm,
+                  OptionalAttr<I32Attr>:$row_shl,
+                  OptionalAttr<I32Attr>:$row_shr,
+                  OptionalAttr<I32Attr>:$row_ror,
+                  OptionalAttr<I32Attr>:$row_mask,
+                  OptionalAttr<I32Attr>:$bank_mask,
+                  OptionalAttr<BoolAttr>:$bound_ctrl,
+                  OptionalAttr<BoolAttr>:$wave_shl,
+                  OptionalAttr<BoolAttr>:$wave_shr,
+                  OptionalAttr<BoolAttr>:$wave_ror,
+                  OptionalAttr<BoolAttr>:$wave_rol,
+                  OptionalAttr<BoolAttr>:$row_mirror,
+                  OptionalAttr<BoolAttr>:$row_half_mirror,
+                  OptionalAttr<I32Attr>:$row_bcast_15,
+                  OptionalAttr<I32Attr>:$row_bcast_31)> {
+  let summary = "AMDGPU DPP operation";
+  let description = [{
+    This operation represents DPP functionality in a GPU program.
+     DPP provides the following operations:  
+  
+    - Full crossbar in a group of four (`quad_perm`)
+    - Wavefront shift left by one lane (`wave_shl`)
+    - Wavefront shift right by one lane (`wave_shr`)
+    - Wavefront rotate right by one lane (`wave_ror`)
+    - Wavefront rotate left by one lane (`wave_rol`)
+    - Row shift left by 1–15 lanes (`row_shl`)
+    - Row shift right by 1–15 lanes (`row_shr`)
+    - Row rotate right by 1–15 lanes (`row_ror`)
+    - Reverse within a row (`row_mirror`)
+    - Reverse within a half-row (`row_half_mirror`)
+    - Broadcast the 15th lane of each row to the next row (`row_bcast`)
+    - Broadcast lane 31 to rows 2 and 3 (`row_bcast`)
+  }];
+  let results = (outs AnyType:$result);
+  let assemblyFormat = [{
+    operands attr-dict `:` type($result)
```

Comment:
Let's give this a slightly nicer assembly format, for instance, assuming a DPP scheme gets added:

```
$src $kind (`(` $permArgument^ `)`)? oilist (`row_mask` $rowMask | `bank_mask` $bankMask | `bound_ctrl` $boundCtrl) `:` type($src)
```